lib/vagrant/plugin/state_file.rb

```diff
@@ -91,9 +92,10 @@ def sources
 
       # This saves the state back into the state file.
       def save!
-        @path.open("w+") do |f|
-          f.write(JSON.dump(@data))
-        end
+        tmpfile = Tempfile.new(@path.basename.to_s, @path.dirname.to_s)
+        tmpfile.write(JSON.dump(@data))
+        tmpfile.close
+        File.rename(tmpfile.path, @path)
```

Comment:
Yea, so there's a reason we don't do this...

1. `File.rename` has historically been a problem on Windows
1. `File.rename` does not work across multiple volumes, particularly if the volumes are different formats
1. The tmpfile could be removed by GC before the mv takes place (aka there's a slight race between lines 97 and 98 here)

I think it's probably a better idea to do something like:

```ruby
Tempfile.open("vagrant-plugins") do |f|
  f.binmode
  f.write(JSON.dump(@data))
  f.close
  FileUtils.mv(f.path, @path)
end
```

What do you think?